app/org/sagebionetworks/bridge/models/ClientInfo.java

```diff
@@ -0,0 +1,272 @@
+
+    private final String appName;
+    private final Integer appVersion;
+    private final String osName;
+    private final String osVersion;
+    private final String sdkName;
+    private final Integer sdkVersion;
+
+    private ClientInfo(String appName, Integer appVersion, String osName, String osVersion, String sdkName,
+            Integer sdkVersion) {
+        this.appName = appName;
+        this.appVersion = appVersion;
+        this.osName = osName;
+        this.osVersion = osVersion;
+        this.sdkName = sdkName;
+        this.sdkVersion = sdkVersion;
+    }
+
+    public String getAppName() {
+        return appName;
+    }
+
+    public Integer getAppVersion() {
+        return appVersion;
+    }
+
+    public String getOsName() {
+        return osName;
+    }
+
+    public String getOsVersion() {
+        return osVersion;
+    }
+
+    public String getSdkName() {
+        return sdkName;
+    }
+
+    public Integer getSdkVersion() {
+        return sdkVersion;
+    }
+    
+    public boolean isTargetedAppVersion(Integer minValue, Integer maxValue) {
+        // If there's no declared client version, it matches anything.
+        if (appVersion != null) {
+            // Otherwise we can't be outside of either range boundary if the boundary
+            // is declared.
+            if ((minValue != null && appVersion.intValue() < minValue.intValue()) || 
+                (maxValue != null && appVersion.intValue() > maxValue.intValue())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + Objects.hashCode(appName);
+        result = prime * result + Objects.hashCode(appVersion);
+        result = prime * result + Objects.hashCode(osName);
+        result = prime * result + Objects.hashCode(osVersion);
+        result = prime * result + Objects.hashCode(sdkName);
+        result = prime * result + Objects.hashCode(sdkVersion);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null || getClass() != obj.getClass())
+            return false;
+        ClientInfo other = (ClientInfo) obj;
+        return Objects.equals(appName, other.appName) && Objects.equals(appVersion, other.appVersion)
+                && Objects.equals(osName, other.osName) && Objects.equals(osVersion, other.osVersion)
+                && Objects.equals(sdkName, other.sdkName) && Objects.equals(sdkVersion, other.sdkVersion);
+    }
+
+    @Override
+    public String toString() {
+        return "ClientInfo [appName=" + appName + ", appVersion=" + appVersion + ", osName=" + osName + ", osVersion="
+                + osVersion + ", sdkName=" + sdkName + ", sdkVersion=" + sdkVersion + "]";
+    }
+
+    public static class Builder {
```

Comment:
Is it worth making the builder package-scoped to force all non-unit-tests to go through the cache?